# L2 Escape Models

The following models take into account that, throughout analyses, *no_goal* seems to behave entirely differently to the other two Goal Type conditions (*goal_frequent* and *goal_non_frequent*). 

Previous focus structure effects (see notebooks 11 to 6) as well as furhter exhaustive modelling (see notebooks 11_appendix vs. 12_appendix) show consistently the heterogeneity of combined data; by spliting data into goal contexts and no-goal contexts, both effects as well as models improved despite the pruning of observations. No-goal contexts are also analysed separately; however, and although results are coherent with what is seen by comparing combined data to goal behaviour data, the sample is clearly not enough to show anything significant from no-goal contexts.

Taking into account the effects summary (notebook 5) and response opportunity analysis (notebook 6), these were chosen to be the best models to predict participants escaping L2-other options.

Import Libraries

In [89]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Goal Behaviour Subgroup

1. Create escape_L2 (goals and no_goals)
2. Counts and Proportions (goals and no_goals)
3. Logistic Models

    - Goal Contexts

        Main Effect Model:

        - E02: Agent

        Additive Models:

        - E04: Focus + Agent
        - E06: Agent + Goal_Type
        - E07: Focus + Agent + Goal_Type

    - No-goal Contexts

        Main Effect Models:

        - E12: Focus
        - E13: Agent

4. Model Comparisons
5. Summary

Read Data

In [90]:
subgroup_theoretical = pd.read_csv("../../data/processed/subgroup_theoretical.csv")

1. Create escape_2

Binary Outcome: responses that are not "L2_other": those who managed to escape L2_other by choosing either a *correct* response or an *L1_transfer* response.

In [91]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

Split data into goals vs. no_goals

In [92]:
goals = subgroup_theoretical[subgroup_theoretical["Goal_Type"]!= "no_goal"].copy()

In [93]:
no_goal = subgroup_theoretical[subgroup_theoretical["Goal_Type"]== "no_goal"].copy()

2. Counts and Proportions for goals and no_goal.

General escape_L2 reponses count:

In [94]:
escape_L2_counts = pd.crosstab(
    subgroup_theoretical["escape_L2"],
    subgroup_theoretical["Response_Full"],
    margins= True)

escape_L2_counts

Response_Full,L1_transfer,L2_other,correct,missing_response,All
escape_L2,,,,,
0,0,38,0,0,38
1,96,0,61,3,160
All,96,38,61,3,198


Escape_L2 responses count for goals:

In [95]:
goal_escape_L2_counts = pd.crosstab(
    goals["escape_L2"],
    goals["Response_Full"],
    margins= True)

goal_escape_L2_counts

Response_Full,L1_transfer,L2_other,correct,missing_response,All
escape_L2,,,,,
0,0,26,0,0,26
1,62,0,43,1,106
All,62,26,43,1,132


Escape_L2 responses count for no_goals:

In [96]:
no_goal_escape_L2_counts = pd.crosstab(
    no_goal["escape_L2"],
    no_goal["Response_Full"],
    margins= True)

no_goal_escape_L2_counts

Response_Full,L1_transfer,L2_other,correct,missing_response,All
escape_L2,,,,,
0,0,12,0,0,12
1,34,0,18,2,54
All,34,12,18,2,66


General escape_L2 responses proportions:

In [97]:
escape_L2_props = subgroup_theoretical["escape_L2"].value_counts(normalize = True)
escape_L2_props

escape_L2
1    0.808081
0    0.191919
Name: proportion, dtype: float64

From 198 observations, 80.8% of responses managed to escape L2_other options. 

Escape_L2 responses proportions for goals:

In [98]:
goals_escape_L2_props = goals["escape_L2"].value_counts(normalize = True)
goals_escape_L2_props

escape_L2
1    0.80303
0    0.19697
Name: proportion, dtype: float64

From 132 observations in the goal group, 80.3% of responses managed to escape L2_other options. 

Escape_L2 responses proportions for no_goals:

In [99]:
no_goal_escape_L2_props = no_goal["escape_L2"].value_counts(normalize = True)
no_goal_escape_L2_props

escape_L2
1    0.818182
0    0.181818
Name: proportion, dtype: float64

From 66 observations in the no_goal group, 81.8% of responses managed to escape L2_other options. 

Counts and Proportions by Condition:

In [100]:
goals_counts_condition = pd.crosstab(
    [goals["Goal_Type"], goals["Agent"], goals["Focus"]],
    goals["escape_L2"],
    margins = True)

goals_counts_condition

escape_L2                       0    1  All
Goal_Type         Agent Focus              
goal_frequent     0     I       8   16   24
                        They    6   12   18
                  1     I       2   14   16
                        They    0    8    8
goal_non_frequent 0     I       8   16   24
                        They    1   17   18
                  1     I       1   15   16
                        They    0    8    8
All                            26  106  132

In [101]:
no_goal_counts_condition = pd.crosstab(
    [no_goal["Agent"],no_goal["Focus"]],
    no_goal["escape_L2"],
    margins = True)

no_goal_counts_condition

escape_L2     0   1  All
Agent Focus             
0     I       3  21   24
      They    5  13   18
1     I       2  14   16
      They    2   6    8
All          12  54   66

In [102]:
goal_props_condition = pd.crosstab(
    [goals["Goal_Type"], goals["Agent"], goals["Focus"]],
    goals["escape_L2"],
    normalize= "index")

goal_props_condition

escape_L2                             0         1
Goal_Type         Agent Focus                    
goal_frequent     0     I      0.333333  0.666667
                        They   0.333333  0.666667
                  1     I      0.125000  0.875000
                        They   0.000000  1.000000
goal_non_frequent 0     I      0.333333  0.666667
                        They   0.055556  0.944444
                  1     I      0.062500  0.937500
                        They   0.000000  1.000000

Within goals, Agency seems to assist in the process of escaping the L2_other options.

In [103]:
no_goal_props_condition = pd.crosstab(
    [no_goal["Agent"],no_goal["Focus"]],
    no_goal["escape_L2"],
    normalize= "index")

no_goal_props_condition

escape_L2           0         1
Agent Focus                    
0     I      0.125000  0.875000
      They   0.277778  0.722222
1     I      0.125000  0.875000
      They   0.250000  0.750000

Within no-goal contexts, however, Agency doesn't seem to play the same part in the process of escaping the L2_other.

On proportions:

- Agency appears to help participants in goal-contexts escape L2. 

- As previously mentioned, if there is a goal representation available, agency seems to help participants commit by leaving the L2 space; however, if there is no goal representation, agency seems to have much less to work with.

3. Logistic Models

Import Libraries

In [104]:
import statsmodels.formula.api as smf
import statsmodels.api as sm
from scipy.stats import chi2


Sanity checks:

In [105]:
goals["escape_L2"].value_counts()

escape_L2
1    106
0     26
Name: count, dtype: int64

In [106]:
no_goal["escape_L2"].value_counts()

escape_L2
1    54
0    12
Name: count, dtype: int64

**Goal Contexts**

**Main Effect Models**

**E02**

escape_L2 ~ Agent

In [107]:
E02 = smf.logit(
    "escape_L2 ~ Agent",
    data=goals
    ).fit()

print(E02.summary())

Optimization terminated successfully.
         Current function value: 0.458568
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  132
Model:                          Logit   Df Residuals:                      130
Method:                           MLE   Df Model:                            1
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.07579
Time:                        13:32:39   Log-Likelihood:                -60.531
converged:                       True   LL-Null:                       -65.495
Covariance Type:            nonrobust   LLR p-value:                  0.001628
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.9754      0.245      3.986      0.000       0.496       1.455
Agent          1.7327      0.

When the data was combined, agency on its own's p-value = 0.013 (coeff = 1.1086). Once the data is the goal split is introduced, agency coefficient increases (1.7327) and p-value improves ( 0.007). Thus, after removing 66 observations, the effect gets actually stronger. 

**Additive Models**

**E04**

escape_L2 ~ Focus + Agent

In [108]:
E04 = smf.logit(
    "escape_L2 ~  Focus + Agent",
    data=goals
    ).fit()

print(E04.summary())

Optimization terminated successfully.
         Current function value: 0.446122
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  132
Model:                          Logit   Df Residuals:                      129
Method:                           MLE   Df Model:                            2
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                  0.1009
Time:                        13:32:39   Log-Likelihood:                -58.888
converged:                       True   LL-Null:                       -65.495
Covariance Type:            nonrobust   LLR p-value:                  0.001351
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         0.6432      0.298      2.156      0.031       0.058       1.228
Focus[T.They]     0.

Focus on its own couldn't predict anything. Again, once it is added to the model with agency, while it is still non-significant (p-value 0 0.080), it is much more informative than boefore. The coefficient also grows from 0.694 to 0.876; so, once agency is accounted for, focus explains additional variation in the data.

The whole model's LLR p-value (0.00135) is also better than agent alone (0.00163); not a lot but not requiring an interaction.

This model explains a lot of variation and demostrates parsimony.


**E06**

escape_L2 ~ Agent + Goal_Type

In [109]:
E06= smf.logit(
    "escape_L2 ~  Agent + Goal_Type",
    data=goals
    ).fit()

print(E06.summary())

Optimization terminated successfully.
         Current function value: 0.451528
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  132
Model:                          Logit   Df Residuals:                      129
Method:                           MLE   Df Model:                            2
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.08998
Time:                        13:32:39   Log-Likelihood:                -59.602
converged:                       True   LL-Null:                       -65.495
Covariance Type:            nonrobust   LLR p-value:                  0.002758
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.6854      0.317      2.162     

This is a good model even though the goal_non_frequent's p is still not siginificant. It is contributing to the model. Given this, a full additive model that was not considered in the beginning should be looked into.

**E07**

escape_L2 ~ Focus + Agent + Goal_Type

In [110]:
E07= smf.logit(
    "escape_L2 ~  Focus + Agent + Goal_Type",
    data=goals
    ).fit()

print(E07.summary())

Optimization terminated successfully.
         Current function value: 0.438864
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                  132
Model:                          Logit   Df Residuals:                      128
Method:                           MLE   Df Model:                            3
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                  0.1155
Time:                        13:32:39   Log-Likelihood:                -57.930
converged:                       True   LL-Null:                       -65.495
Covariance Type:            nonrobust   LLR p-value:                  0.001709
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.3380      0.368      0.918     

Goal Type's p-value improves a little bit and the coefficient increases. The whole model is better (LLR p-value: 0.001709 vs. 0.002758) although not better than the focus-agency additive model's p (0.001351). 

Yet, this model shows a better picture of how counterfactual processing is shaped by considering the 3 cognitive dimensions (Focus, Agency and Goal_Type). In this sense, Agency shows a strong contribution, Focus shows a consistent positive contribution approaching significance and Goal Type shows a modest positive contribution. 

**No-goal Contexts**

**Main Effect Models**

**E12**

escape_L2 ~ Focus

In [111]:
E12 = smf.logit(
    "escape_L2 ~ Focus",
    data=no_goal
    ).fit()

print(E12.summary())

Optimization terminated successfully.
         Current function value: 0.457812
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                   66
Model:                          Logit   Df Residuals:                       64
Method:                           MLE   Df Model:                            1
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.03444
Time:                        13:32:39   Log-Likelihood:                -30.216
converged:                       True   LL-Null:                       -31.293
Covariance Type:            nonrobust   LLR p-value:                    0.1421
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept         1.9459      0.478      4.070      0.000       1.009       2.883
Focus[T.They]    -0.

**E13**

escape_L2 ~ Agent

In [112]:
E13 = smf.logit(
    "escape_L2 ~ Agent",
    data=no_goal
    ).fit()

print(E13.summary())

Optimization terminated successfully.
         Current function value: 0.473694
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:              escape_L2   No. Observations:                   66
Model:                          Logit   Df Residuals:                       64
Method:                           MLE   Df Model:                            1
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:               0.0009392
Time:                        13:32:39   Log-Likelihood:                -31.264
converged:                       True   LL-Null:                       -31.293
Covariance Type:            nonrobust   LLR p-value:                    0.8084
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.4469      0.393      3.682      0.000       0.677       2.217
Agent          0.1625      0.

- While goal participants seem to produce a coherent behaviourla structure, the no_goal participants don't. They don't seem to have the same behaviour constraints. 

- In addition, and even though there are only 66 observations to render anything statistically conclusive, escaping the L2 landscape for no-goal context modelling shows to be coherent with EDA; if anything helped participants escape the L2 landscape, it would be more likely to be Focus (p-value = 0.146) and surely not Agency (p-value = 0.809) -quite the opposite to what happens when looking into goal-context behaviour.

4. Model Comparisons

In [113]:
from scipy.stats import chi2

def lr_test(model_small, model_large):

    LR = 2 * (model_large.llf - model_small.llf)

    df = model_large.df_model - model_small.df_model

    p = chi2.sf(LR, df)

    return pd.Series({
        "LR": LR,
        "df": df,
        "p_value": p
    })

**Unique contributions to additive models**

Agent (E02)

↓

Focus + Agent (E04)

In [114]:
lr_test(E02, E04)

LR         3.285560
df         1.000000
p_value    0.069892
dtype: float64

- Does Focus add information after Agent?

This is week evidence that, once the variability added by Agent is removed, Focus contributes something to escaping the L2. Not to claim it statistically, but enough to have it in the additive model.

Agent (E02)

↓

Agent + Goal_Type (E06)

In [115]:
lr_test(E02, E06)

LR         1.858509
df         1.000000
p_value    0.172797
dtype: float64

- Does Goal_Type add information after Agent?

No evidence that goal type adds any information once the variability added by Agent is removed. 

**Unique contributions to the full additive model**

Focus + Agent (E04)

↓

Focus + Agent + Goal_Type (E07)

In [116]:
lr_test(E04, E07)

LR         1.916139
df         1.000000
p_value    0.166283
dtype: float64

- Does Goal_Type add information after Focus and Agent?

No. Once Focus and Agency are already in the Model, Goal_Type doesn't contribute enough additional information. This confirms that leaving the L2-other alternatives is primarily driven by Agency. 

Agent + Goal_Type (E06)

↓

Focus + Agent + Goal_Type (E07)

In [117]:
lr_test(E06, E07)

LR         3.343190
df         1.000000
p_value    0.067484
dtype: float64

- Does Focus add information after Agent and Goal_Type?

This is week evidence that, once the variability added by no_goals is removed, Focus contributes something to escaping the L2. Not to claim it statistically, but enough to have it in the additive model.

5. Summary

- Agency predicts escaping alternative L2 responses among participants who actually have a goal to protect. Thus, we could ask if ,once there isn't a goal scenario, what could agency help or contribute to. 

- Goal Type consistently exhibited positive coefficients across additive models, although its independent contribution remained modest and did not reach conventional levels of statistical significance.

- Given results of the full additive model, escape responses are best understood as the joint product of agency, perspective, and goal structure, although agency provides the strongest independent contribution. Adding Goal Type preserves the pattern of effects and yields a coherent cognitive account in which agency remains the dominant predictor while focus and goal structure contribute more modestly.

From Model Comparisons:

- There is week eveidence that Focus contributes something to escaping the L2. In contrast, Goal_Type does not significantly improve a model already containing Focus and Agency.

- Separating the no-goal condition reduced heterogeneity in the remaining data, resulting in clearer predictor hierarchies and more interpretable model comparisons.

- Regarding the no-goal subgroup, model comparisons suggested that Focus captured more unique inforamtion than Agency. Whereas the oppostite hierarchy characterised the goal conditions. They are underpowered because of sample size, but the pattern is consistent with the descriptive anaylyses. 